Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -0,0 +1,90 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2020 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// Handles a streaming download to a given file path, allowing headers and progress to be reported.
+public final class FileDownloadDelegate: HTTPClientResponseDelegate {
+    /// The response type for this delegate: the total count of bytes as reported by the response
+    /// "Content-Length" header (if available) and the count of bytes downloaded.
+    public typealias Response = (totalBytes: Int?, receivedBytes: Int)
+
+    private var totalBytes: Int?
+    private var receivedBytes = 0
+
+    private let handle: NIOFileHandle
+    private let io: NonBlockingFileIO
+    private let reportHeaders: ((HTTPHeaders) -> Void)?
+    private let reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)?
+
+    private var writeFuture: EventLoopFuture<Void>?
+
+    /// Initializes a new file download delegate.
+    /// - parameters:
+    ///     - path: Path to a file you'd like to write the download to.
+    ///     - pool: A thread pool to use for asynchronous file I/O.
+    ///     - reportHeaders: A closure called when the response headers are available.
+    ///     - reportProgress: A closure called when a body chunk has been downloaded, with
+    ///       the total byte count and download byte count passed to it as arguments. The callbacks
+    ///       will be invoked in the same threading context that the delegate itself is invoked,
+    ///       as controlled by `EventLoopPreference`.
+    public init(
+        path: String,
+        pool: NIOThreadPool = NIOThreadPool(numberOfThreads: 1),
+        reportHeaders: ((HTTPHeaders) -> Void)? = nil,
+        reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)? = nil
+    ) throws {
+        self.handle = try NIOFileHandle(path: path, mode: .write, flags: .allowFileCreation())
```

Comment:
I know this is annoying, but `open` is technically a blocking operation. We should consider making the handle available by a future using `self.io.openFile(path:mode:flags:)`.

Reply:
Thanks, I've also created https://github.com/apple/swift-nio/pull/1630 to make this more clear in `NIOFaileHandle` doc comments.

Reply:
@Lukasa BTW, there's no such thing as `openFile(path:mode:flags:)`, there's only `openFile(path:mode:flags:eventLoop:)`, where would you like the event loop in the initializer to come from?

Reply:
It'll need to be the event loop that the delegate callbacks fire on.

Reply:
The event loop for delegate callbacks is passed to every callback separately as an argument, do you think there's a way to get access to that event loop in the initializer? Whoever creates the delegate doesn't have access to that event loop too, it's internal and owned by the `HTTPClient` instance, to which the delegate doesn't have any direct access to (or even an indirect one, as far as I understand).

Reply:
Yeah, it's possible that `open` may have to be delayed until the first operation on the file occurs.

Reply:
That's a great idea, I've implemented it and it's now ready for review.